cpp/src/arrow/util/bit-util.cc

```diff
@@ -108,47 +108,85 @@ int64_t CountSetBits(const uint8_t* data, int64_t bit_offset, int64_t length) {
-  uint8_t* dest = buffer->mutable_data();
-
+void TransferBitmap(const uint8_t* data, int64_t offset, int64_t length,
+                    int64_t dest_offset, uint8_t* dest) {
   int64_t byte_offset = offset / 8;
   int64_t bit_offset = offset % 8;
+  int64_t dest_byte_offset = dest_offset / 8;
+  int64_t dest_bit_offset = dest_offset % 8;
   int64_t num_bytes = BitUtil::BytesForBits(length);
-  int64_t bits_to_zero = num_bytes * 8 - length;
+  // Shift dest by its byte offset
+  dest += dest_byte_offset;
 
-  if (bit_offset > 0) {
-    uint32_t carry_mask = BitUtil::kBitmask[bit_offset] - 1U;
-    uint32_t carry_shift = 8U - static_cast<uint32_t>(bit_offset);
-
-    uint32_t carry = 0U;
-    if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
-      carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+  if (dest_bit_offset > 0) {
+    for (int64_t i = 0; i < length; i++) {
+      if (BitUtil::GetBit(data, i + offset)) {
+        BitUtil::SetBit(dest, i + dest_bit_offset);
+      } else {
+        BitUtil::ClearBit(dest, i + dest_bit_offset);
+      }
+    }
+  } else {
+    // Take care of the trailing bits in the last byte
+    int64_t trailing_bits = num_bytes * 8 - length;
+    uint8_t trail = 0;
+    if (trailing_bits) {
+      trail = dest[num_bytes - 1];
     }
 
-    int64_t i = num_bytes - 1;
-    while (i + 1 > 0) {
-      uint8_t cur_byte = data[byte_offset + i];
+    if (bit_offset > 0) {
+      uint32_t carry_mask = BitUtil::kBitmask[bit_offset] - 1U;
+      uint32_t carry_shift = 8U - static_cast<uint32_t>(bit_offset);
+
+      uint32_t carry = 0U;
+      if (BitUtil::BytesForBits(length + bit_offset) > num_bytes) {
+        carry = (data[byte_offset + num_bytes] & carry_mask) << carry_shift;
+      }
+
+      int64_t i = num_bytes - 1;
+      while (i + 1 > 0) {
+        uint8_t cur_byte = data[byte_offset + i];
+        if (invert_bits) {
+          dest[i] = static_cast<uint8_t>(~((cur_byte >> bit_offset) | carry));
+        } else {
+          dest[i] = static_cast<uint8_t>((cur_byte >> bit_offset) | carry);
+        }
+        carry = (cur_byte & carry_mask) << carry_shift;
+        --i;
+      }
+    } else {
       if (invert_bits) {
-        dest[i] = static_cast<uint8_t>(~((cur_byte >> bit_offset) | carry));
+        for (int64_t i = 0; i < num_bytes; i++) {
+          dest[i] = static_cast<uint8_t>(~(data[byte_offset + i]));
+        }
       } else {
-        dest[i] = static_cast<uint8_t>((cur_byte >> bit_offset) | carry);
+        std::memcpy(dest, data + byte_offset, static_cast<size_t>(num_bytes));
       }
-      carry = (cur_byte & carry_mask) << carry_shift;
-      --i;
     }
-  } else {
-    if (invert_bits) {
-      for (int64_t i = 0; i < num_bytes; i++) {
-        dest[i] = static_cast<uint8_t>(~(data[byte_offset + i]));
+
+    for (int i = 0; i < trailing_bits; i++) {
+      if (BitUtil::GetBit(&trail, i + 8 - trailing_bits)) {
+        BitUtil::SetBit(dest, length + i);
```

Comment:
It also seems this fails to take `invert_bits` into account.

Reply:
This is just copying bits that were previously at the end of the bitmap. This is not data that should be modified by the code here.

Reply:
You mean those are padding bits? If so, I think it would be nice to comment a bit more clearly on that :-)
(also, I'm not sure why we would copy those padding bits exactly, instead of simply zeroing them in the destination)

Reply:
Also, you are actually zeroing padding bits in `TransferBitmap(MemoryPool* pool, ...)` below.

Reply:
One of the main changes in this PR is that `CopyBitmap` now supports copying one bitmap into a range of another. This also implies that we could have trailing bits that are already filled. We should not destruct that information (the unit test checks for that).

But correct, in the case that we don't care about the last, we should not execute this path. I'll add a template parameter.